hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/CellBlocks.java

```diff
@@ -0,0 +1,427 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Cellersion 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY CellIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.regionserver;
+
+import org.apache.hadoop.hbase.Cell;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentNavigableMap;
+
+/**
+ * CellBlocks stores a constant number of elements and is immutable after creation stage.
+ * Due to being immutable the CellBlocks can be implemented as array.
+ * The actual array is on- or off-heap and is implemented in concrete class derived from CellBlocks.
+ * The CellBlocks uses no synchronization primitives, it is assumed to be created by a
+ * single thread and then it can be read-only by multiple threads.
+ */
+public abstract class CellBlocks implements ConcurrentNavigableMap<Cell,Cell> {
+
+  private final Comparator<? super Cell> comparator;
+  private int minCellIdx   = 0;   // the index of the minimal cell (for sub-sets)
+  private int maxCellIdx   = 0;   // the index of the maximal cell (for sub-sets)
+  private boolean isDescending = false; // array can be easily traversed backward in descending order
```

Comment:
should isDescending be a property of CellBlocks or only of its iterator? 
Put in other words - we can assume keys in cell block are always in ascending order but you can generate either a forward iterator that traverses them in ascending order or a backward iterator which traverses them in descending order.